BlazarData/src/main/java/com/hubspot/blazar/data/service/StateService.java

```diff
@@ -36,13 +52,40 @@ public StateService(StateDao stateDao, ModuleService moduleService) {
   }
 
   public Set<ModuleState> getModuleStatesByBranch(int branchId) {
+    LOG.debug("Getting composite state for branch {}", branchId);
+    long start = System.currentTimeMillis();
     Set<ModuleState> states = new HashSet<>();
+    List<CompletableFuture<ModuleState>> futureModuleStates = new ArrayList<>();
     for (Module m : moduleService.getByBranch(branchId)) {
-      Optional<ModuleState> state = stateDao.getModuleStateById(m);
-      if (state.isPresent()) {
-        states.add(state.get());
+      futureModuleStates.add(getModuleState(m));
+    }
+    for (CompletableFuture<ModuleState> future : futureModuleStates) {
+      try {
+        states.add(future.get());
+      } catch (InterruptedException e) {
+        LOG.error("Getting module state was for branch {} was interrupted", branchId, e);
+      } catch (ExecutionException e) {
+        LOG.error("Getting module state was for branch {} encountered an exception", branchId, e);
       }
     }
+    metricRegistry.timer(METRIC_FETCH_BRANCH_STATE_TIMER).update(System.currentTimeMillis() - start, TimeUnit.MILLISECONDS);
+    LOG.debug("Got composite state in {}", System.currentTimeMillis() - start);
     return states;
   }
+
+  private CompletableFuture<ModuleState> getModuleState(Module module) {
+    return CompletableFuture.supplyAsync(() -> {
```

Comment:
I think @axiak has a document somewhere about this, but the tl;dr is you don't want anything making network calls like this to be on the common `ForkJoinPool`